silx/io/utils.py

```diff
@@ -190,34 +187,46 @@ def save1D(fname, x, y, xlabel=None, ylabels=None, filetype=None,
     if xlabel is None:
         xlabel = "x"
     if ylabels is None:
-        if len(numpy.array(y).shape) > 1:
+        if numpy.array(y).ndim > 1:
             ylabels = ["y%d" % i for i in range(len(y))]
         else:
             ylabels = ["y"]
     elif isinstance(ylabels, (list, tuple)):
         # if ylabels is provided as a list, every element must
         # be a string
-        ylabels = [ylabels[i] if ylabels[i] is not None else "y%d" % i
-                   for i in range(len(ylabels))]
+        ylabels = [ylabel if isinstance(ylabel, string_types) else "y%d" % i
+                   for ylabel in ylabels]
 
     if filetype.lower() == "spec":
-        y_array = numpy.asarray(y)
-
-        # make sure y_array is a 2D array even for a single curve
-        if len(y_array.shape) == 1:
-            y_array = y_array.reshape(1, y_array.shape[0])
-        elif len(y_array.shape) > 2 or len(y_array.shape) < 1:
-            raise IndexError("y must be a 1D or 2D array")
-
-        # First curve
-        specf = savespec(fname, x, y_array[0], xlabel, ylabels[0], fmt=fmt,
-                         scan_number=1, mode="w", write_file_header=True,
-                         close_file=False)
-        # Other curves
-        for i in range(1, y_array.shape[0]):
-            specf = savespec(specf, x, y_array[i], xlabel, ylabels[i],
-                             fmt=fmt, scan_number=i + 1, mode="w",
-                             write_file_header=False, close_file=False)
+        # Check if we have regular data:
+        ref = len(x)
+        regular = True
+        for one_y in y:
+            regular &= len(one_y) == ref
+        if regular:
+            if isinstance(fmt, (list, tuple)) and len(fmt) < (len(ylabels) + 1):
+                fmt = fmt + [fmt[-1] * (1 + len(ylabels) - len(fmt))]
+            specf = savespec(fname, x, y, xlabel, ylabels, fmt=fmt,
+                     scan_number=1, mode="w", write_file_header=True,
+                     close_file=False)
+        else:
+            y_array = numpy.asarray(y)
+            # make sure y_array is a 2D array even for a single curve
+            if y_array.ndim == 1:
+                y_array.shape = 1, -1
+            elif y_array.ndim not in [1, 2]:
+                raise IndexError("y must be a 1D or 2D array")
```

Comment:
This should not be an IndexError. Could be a ValueError, or just a RuntimeError

Reply:
The error previously raised was already this one ... I did not want to change it.